File: leetcode-js/146-lru-cache.js
class DLinkedNode {
    constructor() {
        this.key = null;
        this.value = null;
        this.pre = null;
        this.post = null;
    }
}

class LRUCache {
    /**
     * @constructor
     */
    constructor(capacity) {
        this.cache = new Map();
        this.count = 0;
        this.capacity = capacity;
        this.head = new DLinkedNode();
        this.head.pre = null;
        this.tail = new DLinkedNode();
        this.tail.post = null;
        this.head.post = this.tail;
        this.tail.pre = this.head;
    }

    /**
     * @param {number} key
     * @returns {number}
     */
    get(key) {
        const node = this.cache.get(key);
        if (!node) {
            return -1;
        }
        this.moveToHead(node);
        return node.value;
    }

    /**
     * @param {number} key
     * @param {number} value
     * @returns {void}
     */
    set(key, value) {
        const node = this.cache.get(key);
        if (!node) {
            const newNode = new DLinkedNode();
            newNode.key = key;
            newNode.value = value;
            this.cache.set(key, newNode);
            this.addNode(newNode);
            this.count += 1;
            if (this.count > this.capacity) {
                const tail = this.popTail();
                this.cache.delete(tail.key);
                this.count -= 1;
            }
        } else {
            node.value = value;
            this.moveToHead(node);
        }
    }

    addNode(node) {
        node.pre = this.head;
        node.post = this.head.post;
        this.head.post.pre = node;
        this.head.post = node;
    }

    removeNode(node) {
        const pre = node.pre;
        const post = node.post;
        pre.post = post;
        post.pre = pre;
    }

    moveToHead(node) {
        this.removeNode(node);
        this.addNode(node);
    }

    popTail() {
        const res = this.tail.pre;
        this.removeNode(res);
        return res;
    }
}

File: leetcode-java/146-lru-cache.java
public class LRUCache {
    public LRUCache(int capacity) {
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }

    private int capacity;
    private int size = 0;
    private Map<Integer, DLinkedNode> map = new HashMap<>();
    private DLinkedNode head = new DLinkedNode(0, 0);
    private DLinkedNode tail = new DLinkedNode(0, 0);

    public int get(int key) {
        if (map.containsKey(key)) {
            DLinkedNode node = map.get(key);
            moveToHead(node);
            return node.value;
        } else {
            return -1;
        }
    }

    public void set(int key, int value) {
        if (map.containsKey(key)) {
            DLinkedNode node = map.get(key);
            node.value = value;
            moveToHead(node);
        } else {
            if (size == capacity) {
                popTail();
                size--;
            }
            DLinkedNode node = new DLinkedNode(key, value);
            add(node);
            map.put(key, node);
            size++;
        }
    }

    private void moveToHead(DLinkedNode node) {
        remove(node);
        add(node);
    }

    private void popTail() {
        DLinkedNode node = map.get(tail.prev.key);
        map.remove(node.key);
        remove(node);
    }

    private void remove(DLinkedNode node) {
        DLinkedNode prev = node.prev;
        prev.next = node.next;
        node.next.prev = prev;
    }

    private void add(DLinkedNode node) {
        node.next = head.next;
        node.next.prev = node;
        head.next = node;
        node.prev = head;
    }

    private class DLinkedNode {
        DLinkedNode(int key, int value) {
            this.key = key;
            this.value = value;
        }

        int key;
        int value;
        DLinkedNode prev = null;
        DLinkedNode next = null;
    }
}
